test/patterns.tests.ts

```diff
@@ -364,28 +364,445 @@ result = (a, b, c, d, e) is
+        Token.Literals.Char("a"),
+        Token.Punctuation.Char.End,
+        Token.Operators.Word.And,
+        Token.Operators.Relational.LessThanOrEqual,
+        Token.Punctuation.Char.Begin,
+        Token.Literals.Char("z"),
+        Token.Punctuation.Char.End,
+        Token.Operators.Word.Or,
+        Token.Operators.Relational.GreaterThanOrEqual,
+        Token.Punctuation.Char.Begin,
+        Token.Literals.Char("A"),
+        Token.Punctuation.Char.End,
+        Token.Operators.Word.And,
+        Token.Operators.Relational.LessThanOrEqual,
+        Token.Punctuation.Char.Begin,
+        Token.Literals.Char("Z"),
+        Token.Punctuation.Char.End,
+        Token.Punctuation.Semicolon,
+      ]);
     });
 
     it("Parenthesized pattern", async () => {
+      const input = Input.InMethod(`
+result = c is
+    (>= 'a' and <= 'z') or
+    (>= 'A' and <= 'Z');
+`
+      );
+
+      const tokens = await tokenize(input);
 
+      tokens.should.deep.equal([
+        Token.Variables.ReadWrite("result"),
+        Token.Operators.Assignment,
+        Token.Variables.ReadWrite("c"),
+        Token.Keywords.Is,
+        Token.Punctuation.OpenParen,
+        Token.Operators.Relational.GreaterThanOrEqual,
+        Token.Punctuation.Char.Begin,
+        Token.Literals.Char("a"),
+        Token.Punctuation.Char.End,
+        Token.Operators.Word.And,
+        Token.Operators.Relational.LessThanOrEqual,
+        Token.Punctuation.Char.Begin,
+        Token.Literals.Char("z"),
+        Token.Punctuation.Char.End,
+        Token.Punctuation.CloseParen,
+        Token.Operators.Word.Or,
+        Token.Punctuation.OpenParen,
+        Token.Operators.Relational.GreaterThanOrEqual,
+        Token.Punctuation.Char.Begin,
+        Token.Literals.Char("A"),
+        Token.Punctuation.Char.End,
+        Token.Operators.Word.And,
+        Token.Operators.Relational.LessThanOrEqual,
+        Token.Punctuation.Char.Begin,
+        Token.Literals.Char("Z"),
+        Token.Punctuation.Char.End,
+        Token.Punctuation.CloseParen,
+        Token.Punctuation.Semicolon,
+      ]);
     });
   });
 
   describe("switch statement", () => {
+    it("supports patterns", async () => {
+      const input = Input.InMethod(`
+switch (a)
+{
+    case string str when str.Length is >3 and not >8: break;
+    case "" or null or 0: break;
+    case { Length: >8 or <=2 }: break;
+    case (first: int, { A: not null, B.C: string }): break;
+    case _: break;
+}
+`
+      );
 
+      const tokens = await tokenize(input);
+
+      tokens.should.deep.equal([
+        Token.Keywords.Control.Switch,
+        Token.Punctuation.OpenParen,
+        Token.Variables.ReadWrite("a"),
+        Token.Punctuation.CloseParen,
+        Token.Punctuation.OpenBrace,
+        Token.Keywords.Control.Case,
+        Token.PrimitiveType.String,
+        Token.Identifiers.LocalName("str"),
+        Token.Keywords.Control.When,
+        Token.Variables.Object("str"),
```

Comment:
`str` is an object when all on one line but below with line breaks becomes a readwrite.

Reply:
`when` clause -> `case-guard` -> `expression` -> `member-access-expression` <- does not handle line breaks well.

And it's impossible to tell whether a variable is followed by a dot and member access on the next line with textmate.

Personally, I don't like how there are so many different variable tokens (local, object, property, parameter) when it's almost impossible to get them correctly, even sometimes without line breaks.
This type of classification is semantics, and is very easily handled by the semantic highlighting (it forces override any textmate scopes - another behaviour I personally dislike).

Anyhow, we don't need to care about this behaviour here. I might make a PR regarding the variable scopes (plus just all scopes in general) later though.